text/0011-validate-requests.md

```diff
@@ -0,0 +1,72 @@
+# Validate input requests
+
+## Related issues and PRs
+
+- Reference Issues:
+- Implementation PR(s):
+
+## Timeline
+
+- Start Date: 2021-06-23
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Given a schema, _validation_ checks that policies conform to the schema, and _schema-based parsing_ checks that entities conform to the schema.
+This RFC proposes to add support for checking that a query conforms to a schema.
+
+## Basic example
+
+Say the schema specifies that an action `readFile` should only be used with principals of type `User` and resources of type `File`.
+This RFC proposes to change the signature of the `Request` constructor to add an optional schema argument that can be used to check  argument types when constructing a `Request`.
+
+Examples (w/ invalid syntax for EUIDs):
+
+```rust
+Request::new(Some(principal), Some(action), Some(resource), context, None); // no schema; works like before
+Request::new(Some(User::"alice"), Some(Action::"readFile"), Some(File::"secret_file.txt"), context, Some(schema)); // ok
+Request::new(Some(User::"alice"), Some(Action::"readFile"), Some(User::"bob"), context, Some(schema)); // error, invalid resource
+```
+
+## Motivation
+
+(Example migrated from #5)
+
+Say that a user wants to allow any `User` to view files in the folder `Folder::"public"`.
+Currently, to achieve the desired behavior, the user should define a schema with the action entity `readFile` that can apply to principals of type `User` and resources of type `File`, and then write the following policy.
+
+```
+permit(principal, action == Action::"readFile", resource in Folder::"public");
+```
+
+Assuming that they pass in inputs where the principal is of type `User` and the resource is of type `File`, this will work as expected.
+However, it could also return "allow" for other principal types like `Folder` or resource types like `User`.
```

Comment:
This feels contrived. Why would I have a `Folder` principal type when using the `readFile` action? Moreover, why does `User` as a resource type matter in this case? Presumably no `User` entities exist in `Folder::"public"`. Is there a way to tweak your "for example" to make the chances of a bug occurring seem more plausible?

Reply:
I guess a more natural mistake in this setup would for the resource to be a `Folder`. In this case, the result of the policy could be "allow" (assuming that folders can be nested). Is this a more convincing use case?

I'll do some thinking to come up with a better example. Open to suggestions in the meantime.

Reply:
I like the `Folder` use-case. Then you'd have to change your example, though, to use `is` for the resource, rather than the principal, right?